File: src/geometry.rs
//! Geometric primitives for camera models and projections.

use crate::matrix::matrix3;
use crate::utils::Float;
use crate::vect::{vect2, vect3};

/// Pinhole camera model with intrinsics and extrinsics, generic over
/// the scalar like the math types; use [`cameraf`] / [`camerad`].
#[allow(non_camel_case_types)]
pub struct camera<T: Float> {
    // Intrinsics
    pub fx: T,
    pub fy: T,
    pub cx: T,
    pub cy: T,
    pub width: u32,
    pub height: u32,
    // Extrinsics: position and orientation in robot frame
    pub camera_pos: vect3<T>,
    /// Rotation from **c**amera frame to **r**obot frame (`mc2r` = M_camera_to_robot).
    pub mc2r: matrix3<T>,
}

/// f32 camera.
#[allow(non_camel_case_types)]
pub type cameraf = camera<f32>;
/// f64 camera.
#[allow(non_camel_case_types)]
pub type camerad = camera<f64>;
/// Legacy alias of [`cameraf`]; new code names the precision.
pub type Camera = cameraf;

impl<T: Float> camera<T> {
    /// Project a 3D point in camera frame to 2D pixel coordinates.
    pub fn project(&self, p_cam: vect3<T>) -> vect2<T> {
        vect2::new(
            self.fx * p_cam.x / p_cam.z + self.cx,
            self.fy * p_cam.y / p_cam.z + self.cy,
        )
    }

    /// Unproject a pixel to a unit direction in camera frame.
    pub fn unproject(&self, px: vect2<T>) -> vect3<T> {
        let dir = vect3::new(
            (px.x - self.cx) / self.fx,
            (px.y - self.cy) / self.fy,
            T::one(),
        );
        dir * (T::one() / dir.norm())
    }

    /// Transform a world point into this camera's frame given robot pose.
    pub fn world_to_camera(&self, p_world: vect3<T>, robot_pos: vect3<T>, mr2w: matrix3<T>) -> vect3<T> {
        let mw2r = mr2w.transpose();
        let mr2c = self.mc2r.transpose();
        let p_robot = mw2r * (p_world - robot_pos);
        mr2c * (p_robot - self.camera_pos)
    }

    /// Unproject pixel to unit direction in robot frame.
    pub fn unproject_to_robot(&self, px: vect2<T>) -> vect3<T> {
        let dir_cam = self.unproject(px);
        self.mc2r * dir_cam
    }

    /// Angular size of one pixel at position `px`, in radians per axis.
    ///
    /// In a pinhole camera, pixel angular size varies across the image:
    ///   size_x = fx / ((px.x - cx)^2 + fx^2)
    ///   size_y = fy / ((py.y - cy)^2 + fy^2)
    /// At image center this equals 1/fx; at edges it is smaller.
    pub fn pixel_angular_size(&self, px: vect2<T>) -> vect2<T> {
        let dx = px.x - self.cx;
        let dy = px.y - self.cy;
        vect2::new(
            self.fx / (dx * dx + self.fx * self.fx),
            self.fy / (dy * dy + self.fy * self.fy),
        )
    }

    /// Check if a pixel coordinate is within the image bounds.
    pub fn is_visible(&self, px: vect2<T>) -> bool {
        px.x >= T::zero() && px.x < T::from(self.width).unwrap()
            && px.y >= T::zero() && px.y < T::from(self.height).unwrap()
    }
}

File: cargo-arael/python/arael/geometry.py
# arael Python geometry: pinhole camera. Mirrors arael's
# src/geometry.rs (intrinsics + extrinsics in robot frame, mc2r =
# camera-to-robot rotation), generic over the scalar like the math
# types: cameraf / camerad, with Camera as a legacy alias of cameraf.

from .math import matrix3d, matrix3f, vect2d, vect2f, vect3d, vect3f


def _make_camera(name, vec2, vec3, mat3):
    class cam:
        """Pinhole camera model with intrinsics and extrinsics."""

        def __init__(self, fx, fy, cx, cy, width, height, camera_pos, mc2r):
            self.fx = float(fx)
            self.fy = float(fy)
            self.cx = float(cx)
            self.cy = float(cy)
            self.width = int(width)
            self.height = int(height)
            self.camera_pos = (camera_pos if isinstance(camera_pos, vec3)
                               else vec3(camera_pos))
            self.mc2r = mc2r if isinstance(mc2r, mat3) else mat3(mc2r)

        def project(self, p_cam):
            """3D point in camera frame -> 2D pixel coordinates."""
            return vec2(self.fx * p_cam.x / p_cam.z + self.cx,
                        self.fy * p_cam.y / p_cam.z + self.cy)

        def unproject(self, px):
            """Pixel -> unit direction in camera frame."""
            d = vec3((px[0] - self.cx) / self.fx,
                     (px[1] - self.cy) / self.fy, 1.0)
            return d * (1.0 / d.norm())

        def world_to_camera(self, p_world, robot_pos, mr2w):
            """World point -> this camera's frame, given the robot pose."""
            p_robot = mr2w.transpose() * (p_world - robot_pos)
            return self.mc2r.transpose() * (p_robot - self.camera_pos)

        def unproject_to_robot(self, px):
            """Pixel -> unit direction in robot frame."""
            return self.mc2r * self.unproject(px)

        def pixel_angular_size(self, px):
            """Angular size of one pixel at px, radians per axis."""
            dx = px[0] - self.cx
            dy = px[1] - self.cy
            return vec2(self.fx / (dx * dx + self.fx * self.fx),
                        self.fy / (dy * dy + self.fy * self.fy))

        def is_visible(self, px):
            """Whether a pixel coordinate is inside the image bounds."""
            return (0.0 <= px[0] < float(self.width)
                    and 0.0 <= px[1] < float(self.height))

    cam.__name__ = name
    cam.__qualname__ = name
    return cam


cameraf = _make_camera("cameraf", vect2f, vect3f, matrix3f)
camerad = _make_camera("camerad", vect2d, vect3d, matrix3d)
# Legacy alias of cameraf; new code names the precision.
Camera = cameraf
